### PR TITLE
Feat: refactor envbinding for adapt policy placement

### PR DESCRIPTION
**Pull request by zxbyoyoyo**
### Description of your changes

Fixes #

I have:

- [x] Read and followed KubeVela's [contribution process](https://github.com/oam-dev/kubevela/blob/master/contribute/create-pull-request.md).
- [ ] [Related Docs](https://github.com/oam-dev/kubevela.io) updated properly. In a new feature or configuration option, an update to the documentation is necessary. 
- [ ] Run `make reviewable` to ensure this PR is ready for review.
- [ ] Added `backport release-x.y` labels to auto-backport this PR if necessary.

### CODE REVIEW

File: pkg/apiserver/rest/usecase/envbinding.go [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+func (c *envBindingUsecaseImpl) CreateEnvBinding(ctx context.Context, app *model.Application, envReq apisv1.CreateApplicationEnvRequest) (*apisv1.EnvBinding, error) {
+	envBinding, err := c.getBindingByEnv(ctx, app, envReq.Name)
+	if err != nil {
+		if !errors.Is(err, datastore.ErrRecordNotExist) {
+			return nil, err
+		}
+	}
+	envBindingModel := convertCreateReqToEnvBindingModel(app, envReq)
+	if envBinding != nil {
+		if err := c.ds.Put(ctx, &envBindingModel); err != nil {
```
Comment (barnettZQG) [inline pkg/apiserver/rest/usecase/envbinding.go:58]: No updates should be made in the creation logic.  If exist, should return `bcode.ErrorEnvbindExist`

File: pkg/apiserver/rest/usecase/envbinding.go [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+
+func (c *envBindingUsecaseImpl) UpdateEnvBinding(ctx context.Context, app *model.Application,
+	envName string, envUpdate apisv1.PutApplicationEnvRequest) (*apisv1.EnvBinding, error) {
+	envBinding, err := c.getBindingByEnv(ctx, app, envName)
+	if err != nil {
+		return nil, err
+	}
+	var envBind model.EnvBinding
+	// update env-binding base
+	for i, env := range app.EnvBinding {
```
Comment (barnettZQG) [inline pkg/apiserver/rest/usecase/envbinding.go:114]: The EnvBinding field in the app model should be removed.

File: pkg/apiserver/rest/usecase/envbinding.go [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+		Description: envBinding.Description,
+		TargetNames: envBinding.TargetNames,
+	}
+	if envBinding.ComponentSelector != nil {
+		re.ComponentSelector = (*apisv1.ComponentSelector)(envBinding.ComponentSelector)
+	}
+	return re, nil
+}
+
+func (c *envBindingUsecaseImpl) DeleteEnvBinding(ctx context.Context, app *model.Application, envName string) error {
```
Comment (barnettZQG) [inline pkg/apiserver/rest/usecase/envbinding.go:151]: Does this implement batch delete?

File: pkg/apiserver/rest/usecase/envbinding.go [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+	}
+	return envBinding
+}
+
+func converEnvbindingModelToBase(envBinding *model.EnvBinding) *apisv1.EnvBindingBase {
+	ebb := &apisv1.EnvBindingBase{
+		Name:              envBinding.Name,
+		Alias:             envBinding.Alias,
+		Description:       envBinding.Description,
+		TargetNames:       envBinding.TargetNames,
```
Comment (barnettZQG) [inline pkg/apiserver/rest/usecase/envbinding.go:187]: Here you can load the alias and description of DeliveryTarget.

File: pkg/apiserver/rest/usecase/application.go [hunk 5]
```diff
@@ -293,6 +272,114 @@ func (c *applicationUsecaseImpl) CreateApplication(ctx context.Context, req apis
...
+func (c *applicationUsecaseImpl) saveApplicationEnvBinding(ctx context.Context, app model.Application, envBindings []*apisv1.EnvBinding) error {
+	for _, envBinding := range envBindings {
+		//1.create env Binding
+		_, err := c.envBindingUsecase.CreateEnvBinding(ctx, &app, apisv1.CreateApplicationEnvRequest{EnvBinding: *envBinding})
+		if err != nil {
+			return err
+		}
+		//2.create env policy
+		//3.create env workflow
+		for _, env := range envBindings {
```
Comment (barnettZQG) [inline pkg/apiserver/rest/usecase/application.go:334]: remove createEnvPolicy and move createEnvWorkflow to CreateEnvBinding function.

File: pkg/apiserver/rest/usecase/application.go [hunk 6]
```diff
@@ -814,6 +860,13 @@ func (c *applicationUsecaseImpl) DeleteApplication(ctx context.Context, app *mod
 		}
 	}
 
+	for _, envBinding := range envBindings {
```
Comment (barnettZQG) [inline pkg/apiserver/rest/usecase/application.go:863]: Call the batch delete method in the `envbindingUsecase` class?

File: pkg/apiserver/rest/webservice/application.go [hunk 7]
```diff
@@ -96,49 +97,6 @@ func (c *applicationWebService) GetWebService() *restful.WebService {
 		Returns(400, "", bcode.Bcode{}).
 		Writes(apis.ApplicationBase{}))
 
-	ws.Route(ws.PUT("/{name}/envs/{envName}").To(c.updateApplicationEnvBinding).
```
Comment (barnettZQG) [inline pkg/apiserver/rest/webservice/application.go:99]: This should not be deleted; envBinding is a child resource of the application.

File: pkg/apiserver/rest/usecase/application.go [hunk 8]
```diff
@@ -293,6 +272,114 @@ func (c *applicationUsecaseImpl) CreateApplication(ctx context.Context, req apis
 	return base, nil
 }
 
+func (c *applicationUsecaseImpl) createEnvPolicy(ctx context.Context, app *model.Application, env *apisv1.EnvBinding) (*model.ApplicationPolicy, error) {
+	policy := &model.ApplicationPolicy{
```
Comment (barnettZQG) [inline pkg/apiserver/rest/usecase/application.go:276]: The policy model does not isolate env, so it cannot be saved to db.

File: pkg/apiserver/rest/usecase/envbinding.go (hunk 4 above)
Comment (zxbyoyoyo) [reply pkg/apiserver/rest/usecase/envbinding.go:268]: The relationship between envbinding and deliveryTarget is one-to-many

Comment (barnettZQG) [reply pkg/apiserver/rest/usecase/envbinding.go:268]: The purpose of loading the alias is to serve the UI.
suggest:
```
targets:
  - name: xx
     alias: xxxx
```

File: pkg/apiserver/model/envbinding.go [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+
+// PrimaryKey return custom primary key
+func (e *EnvBinding) PrimaryKey() string {
+	return fmt.Sprintf("%s-%s", e.AppPrimaryKey, e.Name)
+}
+
+// Index return custom index
+func (e *EnvBinding) Index() map[string]string {
+	index := make(map[string]string)
+	if e.Name != "" {
```
Comment (barnettZQG) [inline pkg/apiserver/model/envbinding.go:50]: the appPrimaryKey field is required.

File: pkg/apiserver/model/workflow.go [hunk 10]
```diff
@@ -76,7 +75,7 @@ func (w *Workflow) Index() map[string]string {
 	if w.EnvName != "" {
 		index["envName"] = w.EnvName
 	}
-	index["default"] = strconv.FormatBool(w.Default)
+	//index["default"] = strconv.FormatBool(w.Default)
```
Comment (barnettZQG) [inline pkg/apiserver/model/workflow.go:78]:
```suggestion
	index["default"] = strconv.FormatBool(w.Default)
```

File: pkg/apiserver/rest/apis/v1/types.go [hunk 11]
```diff
@@ -253,7 +253,6 @@ type ApplicationBase struct {
 	UpdateTime  time.Time         `json:"updateTime"`
 	Icon        string            `json:"icon"`
 	Labels      map[string]string `json:"labels,omitempty"`
-	EnvBinding  EnvBindingList    `json:"envBinding,omitempty"`
```
Comment (barnettZQG) [inline pkg/apiserver/rest/apis/v1/types.go:256]: This field should not change.

File: pkg/apiserver/rest/webservice/application.go [hunk 12]
```diff
@@ -96,49 +99,6 @@ func (c *applicationWebService) GetWebService() *restful.WebService {
...
-		Filter(c.appCheckFilter).
-		Filter(c.envCheckFilter).
-		Param(ws.PathParameter("name", "identifier of the application ").DataType("string")).
-		Param(ws.PathParameter("envName", "identifier of the application ").DataType("string")).
-		Reads(apis.PutApplicationEnvRequest{}).
-		Returns(200, "", apis.EnvBinding{}).
-		Returns(400, "", bcode.Bcode{}).
-		Writes(apis.EnvBinding{}))
-
-	ws.Route(ws.GET("/{name}/envs/{envName}/status").To(c.getApplicationStatus).
```
Comment (barnettZQG) [inline pkg/apiserver/rest/webservice/application.go:111]: It should not be deleted.

File: pkg/apiserver/rest/apis/v1/types.go (hunk 11 above)
Comment (zxbyoyoyo) [reply pkg/apiserver/rest/apis/v1/types.go:256]: add a envlist interface

**Comment by barnettZQG** [comment]
Continue to optimize after merging.

**Pull request merged**